File: 2734/main.go
package main

import "fmt"

func main() {
	fmt.Println(smallestString("cbabc"))
}

func smallestString(s string) string {
	t := []byte(s)
	firstA := findFirstNotA(s)
	if firstA == len(s) {
		t[firstA-1] = 'z'
		return string(t)
	}

	secondA := findSecondA(s, firstA)
	res := []byte{}
	for i, ch := range t {
		if firstA <= i && i < secondA {
			res = append(res, ch - 1)
		} else {
			res = append(res, ch)
		}
	}
	return string(res)
}

func findSecondA(t string, firstA int) int {
	for i := firstA; i < len(t); i++ {
		if t[i] == 'a' {
			return i
		}
	}
	return len(t)
}

func findFirstNotA(t string) int {
	
	for i, ch := range t {
		if ch != 'a' {
			return i
		}
	}
	return len(t)
}

File: 2734/main.rs
struct Solution{}

impl Solution {
    pub fn smallest_string(s: String) -> String {
        let target = 'a';
        let first_a = s.chars().position(|c|c!=target).unwrap_or(s.len());
        if first_a == s.len() {
            return s[0..s.len()-1].to_string() + "z";
        }
        let second_a = Self::find_second_a(&s, first_a);
        let mut res = String::new();
        for (i, c) in s.chars().enumerate() {
            if first_a <= i && i < second_a {
                res.push((c as u8-1) as char);
            } else {
                res.push(c);
            }
        }
        res
    }

    fn find_second_a(s: &String, first_a: usize) -> usize {
        let chars = s.chars().skip(first_a);
        for (index, c) in chars.enumerate() {
            if c == 'a' {
                return index + first_a;
            }
        }
        s.len()
    }
}

fn main() {
    println!("{}", Solution::smallest_string("cbabc".to_string()))
}